Test suite breaks in CI (xunit-tests / win)

### Dafny version

PR 5621

### Code to produce this issue

```dafny
Run CI on https://github.com/dafny-lang/dafny/pull/5621
```


### Command to run and resulting output

```code
Run ##  Note that, for some mysterious reason, --collect doesn't work with the DafnyLanguageServer.Test package
  ##  Note that, for some mysterious reason, --collect doesn't work with the DafnyLanguageServer.Test package
  dotnet coverage collect -o DafnyLanguageServer.Test.coverage dotnet test --no-restore --blame-hang-timeout 360s --logger "console;verbosity=normal" --logger trx Source/DafnyLanguageServer.Test
  shell: C:\Program Files\PowerShell\7\pwsh.EXE -command ". '{0}'"
  env:
    solutionPath: Source/Dafny.sln
    z3BaseUri: https://github.com/dafny-lang/solver-builds/releases/download/snapshot-2024-04-10
    Logging__LogLevel__Microsoft: Debug
    DOTNET_ROOT: C:\Program Files\dotnet
Microsoft (R) Code Coverage Command Line Tool (x64)
Copyright (c) Microsoft Corporation. All rights reserved.

SessionId: aa998506-0257-413a-8cc4-3336579204d0
  Compiling DafnyRuntimeJava to DafnyRuntimeJava/build/libs/DafnyRuntime-4.6.0.jar...
  > Task :compileJava UP-TO-DATE
  > Task :processResources NO-SOURCE
  > Task :classes UP-TO-DATE
  > Task :jar UP-TO-DATE
  > Task :javadoc UP-TO-DATE
  > Task :javadocJar UP-TO-DATE
  > Task :sourcesJar UP-TO-DATE
  > Task :assemble UP-TO-DATE
  > Task :compileTestJava UP-TO-DATE
  > Task :processTestResources NO-SOURCE
  > Task :testClasses UP-TO-DATE
  > Task :test UP-TO-DATE
  > Task :check UP-TO-DATE
  > Task :build UP-TO-DATE
  
  Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.
  
  You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
  
  For more on this, please refer to https://docs.gradle.org/8.4-rc-3/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
  
  BUILD SUCCESSFUL in 842ms
  7 actionable tasks: 7 up-to-date
  DafnyRuntime -> D:\a\dafny\dafny\dafny\Binaries\netstandard2.0\DafnyRuntime.dll
  DafnyCore -> D:\a\dafny\dafny\dafny\Binaries\net6.0\DafnyCore.dll
  DafnyCore.Test -> D:\a\dafny\dafny\dafny\Source\DafnyCore.Test\bin\Debug\net6.0\DafnyCore.Test.dll
  DafnyPipeline -> D:\a\dafny\dafny\dafny\Binaries\DafnyPipeline.dll
  DafnyLanguageServer -> D:\a\dafny\dafny\dafny\Binaries\DafnyLanguageServer.dll
  The package DafnyPipeline.4.7.0 is missing a readme. Go to https://aka.ms/nuget/authoring-best-practices/readme to learn why package readmes are important.
  Successfully created package 'D:\a\dafny\dafny\dafny\Binaries\DafnyPipeline.4.7.0.nupkg'.
  DafnyLanguageServer.Test -> D:\a\dafny\dafny\dafny\Source\DafnyLanguageServer.Test\bin\Debug\net6.0\DafnyLanguageServer.Test.dll
Test run for D:\a\dafny\dafny\dafny\Source\DafnyLanguageServer.Test\bin\Debug\net6.0\DafnyLanguageServer.Test.dll (.NETCoreApp,Version=v6.0)
VSTest version 17.11.0 (x64)

Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.4.3+1b45f5407b (64-bit .NET 6.0.33)
[xUnit.net 00:00:02.66]   Discovering: DafnyLanguageServer.Test
[xUnit.net 00:00:02.77]   Discovered:  DafnyLanguageServer.Test
[xUnit.net 00:00:02.77]   Starting:    DafnyLanguageServer.Test
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Unit.CompilationManagerTest.CancelUnstartedCompilationLeadsToCancelledTasks [123 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Refinement.RefinementTests.RefinedLemmaNewEnsuresIsVerified [2 s]
fail: Microsoft.Dafny.Compilation[0]
      Caught error in statusUpdates observable.
      System.Exception: CheckerPool was already disposed
         at VC.CheckerPool.FindCheckerFor(Program program, Split split, CancellationToken cancellationToken)
         at Microsoft.Boogie.VerificationTask.StartRun(CancellationToken cancellationToken)+MoveNext()
         at Microsoft.Boogie.VerificationTask.StartRun(CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
         at System.Linq.AsyncEnumerable.ToObservableObservable`1.<>c__DisplayClass2_0.<<Subscribe>g__Core|0>d.MoveNext() in /_/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/ToObservable.cs:line 50
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.MatchExprAndMethodWithoutBody [320 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.DefinitionOfFunctionInvocationOfFunctionDeclaredInForeignDocumentReturnsLocation [145 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.ResolutionErrorInDependent [64 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.DefinitionReturnsBeforeVerificationIsComplete [51 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.JumpToExternModule [134 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.DefinitionOfVariableShadowingFieldReturnsTheFieldIfThisIsUsed [85 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.DefinitionOfInvocationOfUnknownFunctionOrMethodReturnsNoLocation [51 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.ResolutionError [47 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.DefinitionOfVariableShadowedByAnotherReturnsTheOriginalVariable [69 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.DefinitionOfVariableShadowingAnotherVariableReturnsTheShadowingVariable [65 ms]
failfail: Microsoft.Dafny.Compilation[0]
      Caught error in statusUpdates observable.
      System.Exception: CheckerPool was already disposed
         at VC.CheckerPool.FindCheckerFor(Program program, Split split, CancellationToken cancellationToken)
         at Microsoft.Boogie.VerificationTask.StartRun(CancellationToken cancellationToken)+MoveNext()
         at Microsoft.Boogie.VerificationTask.StartRun(CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
         at System.Linq.AsyncEnumerable.ToObservableObservable`1.<>c__DisplayClass2_0.<<Subscribe>g__Core|0>d.MoveNext() in /_/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/ToObservable.cs:line 50
: Microsoft.Dafny.Compilation[0]
      Caught error in statusUpdates observable.
      System.Exception: CheckerPool was already disposed
         at VC.CheckerPool.FindCheckerFor(Program program, Split split, CancellationToken cancellationToken)
         at Microsoft.Boogie.VerificationTask.StartRun(CancellationToken cancellationToken)+MoveNext()
         at Microsoft.Boogie.VerificationTask.StartRun(CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
         at System.Linq.AsyncEnumerable.ToObservableObservable`1.<>c__DisplayClass2_0.<<Subscribe>g__Core|0>d.MoveNext() in /_/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/ToObservable.cs:line 50
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.ResolutionErrorInDependency [77 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.DefinitionOfVariableShadowingFieldReturnsTheVariable [61 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.DefinitionOfFieldOfSystemTypeReturnsNoLocation [68 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.ModuleImport1 [59 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.DefinitionOfMethodInvocationOfMethodDeclaredInSameDocumentReturnsLocation [82 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.Refinement [105 ms]
fail: Microsoft.Dafny.Compilation[0]
      Caught error in statusUpdates observable.
      System.Exception: CheckerPool was already disposed
         at VC.CheckerPool.FindCheckerFor(Program program, Split split, CancellationToken cancellationToken)
         at Microsoft.Boogie.VerificationTask.StartRun(CancellationToken cancellationToken)+MoveNext()
         at Microsoft.Boogie.VerificationTask.StartRun(CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
         at System.Linq.AsyncEnumerable.ToObservableObservable`1.<>c__DisplayClass2_0.<<Subscribe>g__Core|0>d.MoveNext() in /_/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/ToObservable.cs:line 50
fail: Microsoft.Dafny.Compilation[0]
      Caught error in statusUpdates observable.
      System.Exception: CheckerPool was already disposed
         at VC.CheckerPool.FindCheckerFor(Program program, Split split, CancellationToken cancellationToken)
         at Microsoft.Boogie.VerificationTask.StartRun(CancellationToken cancellationToken)+MoveNext()
         at Microsoft.Boogie.VerificationTask.StartRun(CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
         at System.Linq.AsyncEnumerable.ToObservableObservable`1.<>c__DisplayClass2_0.<<Subscribe>g__Core|0>d.MoveNext() in /_/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/ToObservable.cs:line 50
fail: Microsoft.Dafny.Compilation[0]
      Caught error in statusUpdates observable.
      System.Exception: CheckerPool was already disposed
         at VC.CheckerPool.FindCheckerFor(Program program, Split split, CancellationToken cancellationToken)
         at Microsoft.Boogie.VerificationTask.StartRun(CancellationToken cancellationToken)+MoveNext()
         at Microsoft.Boogie.VerificationTask.StartRun(CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
         at System.Linq.AsyncEnumerable.ToObservableObservable`1.<>c__DisplayClass2_0.<<Subscribe>g__Core|0>d.MoveNext() in /_/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/ToObservable.cs:line 50
fail: Microsoft.Dafny.Compilation[0]
      Caught error in statusUpdates observable.
      System.Exception: CheckerPool was already disposed
         at VC.CheckerPool.FindCheckerFor(Program program, Split split, CancellationToken cancellationToken)
         at Microsoft.Boogie.VerificationTask.StartRun(CancellationToken cancellationToken)+MoveNext()
         at Microsoft.Boogie.VerificationTask.StartRun(CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
         at System.Linq.AsyncEnumerable.ToObservableObservable`1.<>c__DisplayClass2_0.<<Subscribe>g__Core|0>d.MoveNext() in /_/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/ToObservable.cs:line 50
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.DatatypesAndMatches [84 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.ModuleImport2 [50 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.DefinitionInConstructorInvocationOfUserDefinedTypeOfForeignFileReturnsLinkToForeignFile [54 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.FunctionCallAndGotoOnDeclaration [84 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.ExplicitProjectToGoDefinitionWorks [105 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.StaticFunctionCall [83 ms]
fail: Microsoft.Dafny.Compilation[0]
      Caught error in statusUpdates observable.
      System.Exception: CheckerPool was already disposed
         at VC.CheckerPool.FindCheckerFor(Program program, Split split, CancellationToken cancellationToken)
         at Microsoft.Boogie.VerificationTask.StartRun(CancellationToken cancellationToken)+MoveNext()
         at Microsoft.Boogie.VerificationTask.StartRun(CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
         at System.Linq.AsyncEnumerable.ToObservableObservable`1.<>c__DisplayClass2_0.<<Subscribe>g__Core|0>d.MoveNext() in /_/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/ToObservable.cs:line 50
Unhandled exception. System.InvalidOperationException: There is no currently active test.
   at Xunit.Sdk.TestOutputHelper.GuardInitialized() in /_/src/xunit.execution/Sdk/Frameworks/TestOutputHelper.cs:line 51
   at Xunit.Sdk.TestOutputHelper.QueueTestOutput(String output) in /_/src/xunit.execution/Sdk/Frameworks/TestOutputHelper.cs:line 60
   at Xunit.Sdk.TestOutputHelper.WriteLine(String message) in /_/src/xunit.execution/Sdk/Frameworks/TestOutputHelper.cs:line 110
   at DafnyCore.Test.WriterFromOutputHelper.WriteLine(String value) in D:\a\dafny\dafny\dafny\Source\DafnyCore.Test\WriterFromOutputHelper.cs:line 28
   at Microsoft.Boogie.SMTLib.SMTLibProcessTheoremProver.HandleProverError(String s)
   at Microsoft.Boogie.SMTLib.SMTLibProcess.HandleError(String msg)
   at Microsoft.Boogie.SMTLib.SMTLibProcess.SolverErrorDataReceived(Object sender, DataReceivedEventArgs e)
   at System.Diagnostics.Process.ErrorReadNotifyUser(String data)
   at System.Diagnostics.AsyncStreamReader.FlushMessageQueue(Boolean rethrowInNewThread)
--- End of stack trace from previous location ---
   at System.Diagnostics.AsyncStreamReader.<>c.<FlushMessageQueue>b__18_0(Object edi)
   at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(QueueUserWorkItemCallback quwi)
   at System.Threading.ExecutionContext.RunForThreadPoolUnsafe[TState](ExecutionContext executionContext, Action`1 callback, TState& state)
   at System.Threading.QueueUserWorkItemCallback.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.ModuleImport1ResolutionErrorInDependency [42 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.WhileLoop [70 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Lookup.GoToDefinitionTest.JumpToOtherModule [112 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Util.StringifyTest.StringifyThing [3 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Synchronization.SymbolMigrationTest.ChangeToSyntacticallyIncorrectDocumentUsesMigration [106 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Synchronization.SymbolMigrationTest.ChangeToSemanticallyCorrectDocumentUsesDafnyResolver [53 ms]
  Passed Microsoft.Dafny.LanguageServer.IntegrationTest.Synchronization.SymbolMigrationTest.ChangeToSemanticallyIncorrectDocumentUsesMigration [101 ms]
The active test run was aborted. Reason: Test host process crashed : Unhandled exception. System.InvalidOperationException: There is no currently active test.
   at Xunit.Sdk.TestOutputHelper.GuardInitialized() in /_/src/xunit.execution/Sdk/Frameworks/TestOutputHelper.cs:line 51
   at Xunit.Sdk.TestOutputHelper.QueueTestOutput(String output) in /_/src/xunit.execution/Sdk/Frameworks/TestOutputHelper.cs:line 60
   at Xunit.Sdk.TestOutputHelper.WriteLine(String message) in /_/src/xunit.execution/Sdk/Frameworks/TestOutputHelper.cs:line 110
   at DafnyCore.Test.WriterFromOutputHelper.WriteLine(String value) in D:\a\dafny\dafny\dafny\Source\DafnyCore.Test\WriterFromOutputHelper.cs:line 28
   at Microsoft.Boogie.SMTLib.SMTLibProcessTheoremProver.HandleProverError(String s)
   at Microsoft.Boogie.SMTLib.SMTLibProcess.HandleError(String msg)
   at Microsoft.Boogie.SMTLib.SMTLibProcess.SolverErrorDataReceived(Object sender, DataReceivedEventArgs e)
   at System.Diagnostics.Process.ErrorReadNotifyUser(String data)
   at System.Diagnostics.AsyncStreamReader.FlushMessageQueue(Boolean rethrowInNewThread)
--- End of stack trace from previous location ---
   at System.Diagnostics.AsyncStreamReader.<>c.<FlushMessageQueue>b__18_0(Object edi)
   at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(QueueUserWorkItemCallback quwi)
   at System.Threading.ExecutionContext.RunForThreadPoolUnsafe[TState](ExecutionContext executionContext, Action`1 callback, TState& state)
   at System.Threading.QueueUserWorkItemCallback.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()


Test Run Aborted.
Total tests: Unknown
     Passed: 31
 Total time: 19.8832 Seconds
Results File: D:\a\dafny\dafny\dafny\Source\DafnyLanguageServer.Test\TestResults\runneradmin_fv-az1457-142_2024-08-21_16_25_56.trx

The active Test Run was aborted because the host process exited unexpectedly. Please inspect the call stack above, if available, to get more information about where the exception originated from.
The test running when the crash occurred: 
Microsoft.Dafny.LanguageServer.IntegrationTest.Synchronization.SymbolMigrationTest.ChangeToDocumentWithVerificationErrorsUsesDafnyResolver

This test may, or may not be the source of the crash.

Attachments:
  D:\a\dafny\dafny\dafny\Source\DafnyLanguageServer.Test\TestResults\1ef6711e-bd98-46fd-a5c4-56041b766e92\Sequence_62dbbb90c64e4e9985a6af485376c959.xml
Code coverage results: DafnyLanguageServer.Test.coverage.
Error: Process completed with exit code 1.
```


### What happened?

CI crashed.


### What type of operating system are you experiencing the problem on?

Other

--- a/Source/DafnyCore.Test/WriterFromOutputHelper.cs
+++ b/Source/DafnyCore.Test/WriterFromOutputHelper.cs
@@ -25,17 +25,29 @@ public class WriterFromOutputHelper : TextWriter {
   public override Encoding Encoding => Encoding.Default;
 
   public override void WriteLine(string? value) {
-    output.WriteLine(buffer + value);
+    try {
+      output.WriteLine(buffer + value);
+    } catch {
+      // ignored
+    }
     buffer.Clear();
   }
 
   public override void WriteLine(string format, params object?[] arg) {
-    output.WriteLine(buffer + format, arg);
+    try {
+      output.WriteLine(buffer + format, arg);
+    } catch {
+      // ignored
+    }
     buffer.Clear();
   }
 
   public override void Flush() {
-    output.WriteLine(buffer.ToString());
+    try {
+      output.WriteLine(buffer.ToString());
+    } catch {
+      // ignored
+    }
     base.Flush();
   }
 }

--- a/Source/DafnyCore.Test/WriterFromOutputHelper.cs
+++ b/Source/DafnyCore.Test/WriterFromOutputHelper.cs
@@ -28,7 +28,7 @@ public class WriterFromOutputHelper : TextWriter {
     try {
       output.WriteLine(buffer + value);
     } catch {
-      // ignored
+      // ignored because of https://github.com/dafny-lang/dafny/issues/5723
     }
     buffer.Clear();
   }
@@ -37,7 +37,7 @@ public class WriterFromOutputHelper : TextWriter {
     try {
       output.WriteLine(buffer + format, arg);
     } catch {
-      // ignored
+      // ignored because of https://github.com/dafny-lang/dafny/issues/5723
     }
     buffer.Clear();
   }
@@ -46,7 +46,7 @@ public class WriterFromOutputHelper : TextWriter {
     try {
       output.WriteLine(buffer.ToString());
     } catch {
-      // ignored
+      // ignored because of https://github.com/dafny-lang/dafny/issues/5723
     }
     base.Flush();
   }

